rust/arrow/src/compute/array_ops.rs

```diff
@@ -204,6 +209,101 @@ where
     Ok(b.finish())
 }
 
+macro_rules! filter_array {
+    ($array:expr, $filter:expr, $array_type:ident) => {{
+        let b = $array.as_any().downcast_ref::<$array_type>().unwrap();
+        let mut builder = $array_type::builder(b.len());
+        for i in 0..b.len() {
+            if $filter.value(i) {
+                builder.append_value(b.value(i))?;
+            }
+        }
+        Ok(Arc::new(builder.finish()))
+    }};
+}
+
+pub fn filter(array: &Array, filter: &BooleanArray) -> Result<ArrayRef> {
+    match array.data_type() {
+        DataType::UInt8 => filter_array!(array, filter, UInt8Array),
+        DataType::UInt16 => filter_array!(array, filter, UInt16Array),
+        DataType::UInt32 => filter_array!(array, filter, UInt32Array),
+        DataType::UInt64 => filter_array!(array, filter, UInt64Array),
+        DataType::Int8 => filter_array!(array, filter, Int8Array),
+        DataType::Int16 => filter_array!(array, filter, Int16Array),
+        DataType::Int32 => filter_array!(array, filter, Int32Array),
+        DataType::Int64 => filter_array!(array, filter, Int64Array),
+        DataType::Float32 => filter_array!(array, filter, Float32Array),
+        DataType::Float64 => filter_array!(array, filter, Float64Array),
+        DataType::Boolean => filter_array!(array, filter, BooleanArray),
+        DataType::Utf8 => {
+            //TODO: this is inefficient and we should improve the Arrow impl to help make
+            // this more concise
+            let b = array.as_any().downcast_ref::<BinaryArray>().unwrap();
+            let mut values: Vec<String> = Vec::with_capacity(b.len());
+            for i in 0..b.len() {
+                if filter.value(i) {
+                    values.push(b.get_string(i));
+                }
+            }
+            let tmp: Vec<&str> = values.iter().map(|s| s.as_str()).collect();
+            Ok(Arc::new(BinaryArray::from(tmp)))
+        }
+        other => Err(ArrowError::ComputeError(format!(
+            "filter not supported for {:?}",
+            other
+        ))),
+    }
+}
+
+macro_rules! limit_array {
+    ($array:expr, $num_rows_to_read:expr, $array_type:ident) => {{
+        let b = $array.as_any().downcast_ref::<$array_type>().unwrap();
+        let mut builder = $array_type::builder($num_rows_to_read);
+        for i in 0..$num_rows_to_read {
+            builder.append_value(b.value(i))?;
+        }
+        Ok(Arc::new(builder.finish()))
+    }};
+}
+
+pub fn limit(array: &Array, num_rows_to_read: usize) -> Result<ArrayRef> {
+    if array.len() < num_rows_to_read {
```

Comment:
I'm curious as to why we return an error here. Why not set the limit to the lower of the 2 instead?

`vec![1,2,3,4,5].iter().take(7)` doesn't throw an error, and `SELECT * FROM table_with_100_records LIMIT 200` will also only return the 100 records. What do you think about the suggestion?

Reply:
My initial thought was that if someone calls this function with a `num_rows_to_read` larger than `array.len()`, they will probably want to know rather than silently choosing the minimum.

However given `take` and `LIMIT` behavior this is indeed probably not what is expected from this function. I will change it, good call.

Reply:
I think if you limit the `num_rows_to_read` to the array length.
Have a look at https://issues.apache.org/jira/browse/ARROW-3954, which will make filter more efficient in future.